actioncable/lib/action_cable/subscription_adapter/redis.rb

```diff
@@ -150,6 +150,9 @@ def ensure_listener_running
 
                 conn = @adapter.redis_connection_for_subscriptions
                 listen conn
+              rescue ::Redis::BaseConnectionError
+                @thread = @raw_client = nil
+                ActionCable.server.restart
```

Comment:
1. While this is an improvement over a wedged process, is there are more graceful strategy than disconnecting all clients?
2. Should this implementation live here, or in a separate method dedicated to handling a failed listener?
3. Should this error be reported so it isn't invisibly handled?

Reply:
>While this is an improvement over a wedged process, is there are more graceful strategy than disconnecting all clients?

This is difficult. Ideally we could send a disconnect message to the **specific connection** that fails with this exception by calling [ApplicationCable::Connection#close](https://github.com/rails/rails/blob/ce9d2a1160af8b726f2e8c8156dad2578d24886d/actioncable/lib/action_cable/connection/base.rb#L106-L114) the same way [ActionCable::Server#restart](https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/actioncable/lib/action_cable/server/base.rb#L41-L44) does for each connection. 

The problem I'm having is identifying the connection within `ActionCable::SubscriptionAdapter::Redis`, we have the "channel" (e.g `action_cable/Z2lkOi8vaGVscC1odWIvVXNlci81MjgyMjY0`) a connection uses but I can't seem to use that to identify the connection and call `#close`. There is code to [identify `RemoteConnections` via identifiers](https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/actioncable/lib/action_cable/remote_connections.rb#L17) but that's not helpful here as it uses custom identifiers added with `identified_by`.

We could also try retrying the Redis connection, although that should happen automatically when clients attempt to resubscribe.

If you have a suggestion around how try and send a disconnect message to the specific client connection that failed, let me know. 


>Should this implementation live here, or in a separate method dedicated to handling a failed listener?

Splitting this out into an isolated method is a good idea, especially if we add connection specific disconnect logic mentioned above.

>Should this error be reported so it isn't invisibly handled?

Agreed, it's likely we should report this error, either by re-raising or logging to avoid this being silenced.

Reply:
I don't think we can disconnect a specific client connection because all those share the same Redis connection.

Just reconnecting also does not work because once the connection fails, we lose the subscriptions that we already made on the connection. So the easiest way is to just disconnect the clients so that they connect again and they send the subscriptions they need.

Reply:
@jeremy Can you clarify if the approach of disconnecting all clients via a server restart is acceptable? As mentioned by @engwan this seems like the best way forward. 

If that's acceptable I'll be happy to spin this out into an isolated method and handle/log the exception.

Reply:
Hey everyone,

> is there are more graceful strategy than disconnecting all clients?

I think, there is. I don't think restarting the whole server (and disconnecting clients) is a good solution. The pub/sub adapter should not _leak_ its failure/restart details to other Action Cable components. Connection/channels instances do not depend on Redis directly.

How can we do that?

We can keep track of channels (actually, streams) and resubscribe when connection is restored.

A straightforward implementation could be to re-use `#add_channel` and `#remove_channel` from the SubsriberMap (which we extends in the Redis adapter) and then use `@subscribers.keys` as the list of current streams. On disconnect, we should add something like `when_connected { subscriber_maps.subscribers.keys.each { @subscribed_client.subscribe(_1) }`.